### PR TITLE
Add PostgreSQL as a subchart to simplify out-of-the-box installation

### LINKED ISSUE #27: Include postgres as an optional sub-chart to make it easier to install the server
probably using the bitnami postgresql chart

see https://github.com/chgl/charts/blob/master/charts/hapi-fhir-jpaserver/Chart.yaml#L10 for an example

### PR DESCRIPTION
**Pull request by chgl**
Closes #27 

This simplifies some parts of the chart using several templates inside the `_helpers.tpl`

### CODE REVIEW

File: charts/ibm-fhir-server/values.yaml [hunk 1]
```diff
@@ -41,6 +41,32 @@ endpoints:
...
       - patch
       - delete
+
+# see <https://github.com/bitnami/charts/tree/master/bitnami/postgresql> for details
+postgresql:
+  # -- enable an included PostgreSQL DB.
+  # if set to `false`, the connection settings under the `db` key are used
+  enabled: true
+  image:
+    # -- update the default Postgres version to 13.4
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/values.yaml:51]: I think this should document the property, not its value.
```suggestion
    # -- the tag for the postgresql image
```

File: charts/ibm-fhir-server/values.yaml [hunk 2]
```diff
@@ -41,6 +41,32 @@ endpoints:
...
+  # -- name of the database to create.
+  # see: <https://github.com/bitnami/bitnami-docker-postgresql/blob/master/README.md#creating-a-database-on-first-run>
+  postgresqlDatabase: "fhir"
+  # -- Name of existing secret to use for PostgreSQL passwords.
+  # The secret has to contain the keys `postgresql-password`
+  # which is the password for `postgresqlUsername` when it is
+  # different of `postgres`, `postgresql-postgres-password` which
+  # will override `postgresqlPassword`, `postgresql-replication-password`
+  # which will override `replication.password` and `postgresql-ldap-password`
+  # which will be sed to authenticate on LDAP. The value is evaluated as a template.
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/values.yaml:62]: I found this section a little hard to parse.  See if this sounds better to you too:
```suggestion
  # -- Name of existing secret to use for PostgreSQL passwords.
  # The secret must contain the keys `postgresql-password`
  # (the password for `postgresqlUsername` when it is different from `postgres`),
  # `postgresql-postgres-password` (which will override `postgresqlPassword`),
  # `postgresql-replication-password` (which will override `replication.password`), and
  # `postgresql-ldap-password` (used to authenticate on LDAP).
  # The value is evaluated as a template.
```

File: charts/ibm-fhir-server/templates/deployment.yaml [hunk 3]
```diff
@@ -27,6 +27,36 @@ spec:
       imagePullSecrets:
         {{- toYaml . | nindent 8 }}
       {{- end }}
+      {{- if eq .Values.db.type "postgresql" }}
+      initContainers:
+        - name: wait-for-db-to-be-ready
+          image: docker.io/bitnami/postgresql:13.4.0
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/templates/deployment.yaml:33]: should we use `postgresql.image.tag` to be consistent with the db image (when it is used)?

File: charts/ibm-fhir-server/Chart.yaml [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
 description: Helm chart for the IBM FHIR Server
 name: ibm-fhir-server
-version: 0.1.2
+version: 0.2.0
 appVersion: 4.9.1
+dependencies:
+  - name: postgresql
+    version: 10.7.0
+    repository: https://charts.bitnami.com/bitnami
+    condition: postgresql.enabled
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/Chart.yaml:10]: we need `postgresqlExtendedConf.maxPreparedTransactions` set to a non-default, non-zero value in order for our bulk data import/export stuff to work.  is there a good way to override the subchart's default from our chart?

**Comment by lmsurpre** [comment]
Thanks @chgl !  I left a few comments but I think the biggest question is around how well the init container approach will work for us.
A nice property of the helm hooks is that it fails the upgrade/install if the create/update schema hooks fail.
Conversely, with this new approach the new server version will start up in parallel with the schema update.
I think that is probably acceptable (and seems better than having all the deployments run a schema update every time they try to start).

We could potentially address that by making a SQL call to the FHIR_ADMIN VERSION_HISTORY table and ensuring that the version matches what we expect before starting up the server, but I don't think it will be an issue unless we make some drastic schema change.

The other thing worth noting is that the current init container approach here is only implemented for postgresql.  We'd need to do something similar for any other db that we want to support (currently just db2) in a similar way.  However, since its much less likely that someone will be adding a subchart for db2 any time soon, that seems fine to me.

File: charts/ibm-fhir-server/templates/deployment.yaml [hunk 5]
```diff
@@ -65,44 +95,31 @@ spec:
...
-            {{- else }}
-            {{- if .Values.db.password }}
-            - name: FHIR_DB_PASSWORD
-              valueFrom:
-                secretKeyRef:
-                  name: {{ template "fhir.fullname" . }}-db-secret
-                  key: password
-            {{- end }}
-            {{- if .Values.db.apiKey }}
+            {{- else if .Values.db.apiKey }}
```
Comment (michaelwschroeder) [inline charts/ibm-fhir-server/templates/deployment.yaml:116]: lines 110-122 can be replaced with the following (based on comment in _helpers.tpl):
```
            - name: FHIR_DB_APIKEY
              valueFrom:
                secretKeyRef:
                  name: "{{ include "fhir.database.secretName" $ }}"
                  key: "{{ include "fhir.database.apiKeySecretKey" $ }}"
```

File: charts/ibm-fhir-server/templates/deployment.yaml [hunk 6]
```diff
@@ -65,44 +95,31 @@ spec:
...
-            {{- if .Values.db.passwordSecretKey }}
+              value: "{{ include "fhir.database.user" $ }}"
             - name: FHIR_DB_PASSWORD
               valueFrom:
                 secretKeyRef:
-                  name: {{ .Values.db.dbSecret }}
-                  key: {{ .Values.db.passwordSecretKey }}
-            {{- end }}
+                  name: "{{ include "fhir.database.secretName" $ }}"
+                  key: "{{ include "fhir.database.secretKey" $ }}"
```
Comment (michaelwschroeder) [inline charts/ibm-fhir-server/templates/deployment.yaml:109]:
```suggestion
                  key: "{{ include "fhir.database.passwordSecretKey" $ }}"
```

per suggestion in _helpers.tpl

File: charts/ibm-fhir-server/templates/deployment.yaml [hunk 7]
```diff
@@ -123,44 +140,32 @@ spec:
...
-            {{- if .Values.db.passwordSecretKey }}
+              value: "{{ include "fhir.database.user" $ }}"
             - name: BATCH_DB_PASSWORD
               valueFrom:
                 secretKeyRef:
-                  name: {{ .Values.db.dbSecret }}
-                  key: {{ .Values.db.passwordSecretKey }}
-            {{- end }}
+                  name: "{{ include "fhir.database.secretName" $ }}"
+                  key: "{{ include "fhir.database.secretKey" $ }}"
```
Comment (michaelwschroeder) [inline charts/ibm-fhir-server/templates/deployment.yaml:154]:
```suggestion
                  key: "{{ include "fhir.database.passwordSecretKey" $ }}"
```
per comment in _helpers.tpl

Comment (michaelwschroeder) [reply charts/ibm-fhir-server/templates/deployment.yaml:154]: lines 155-168 can be replaced with the following ((based on comment in _helpers.tpl):
```
            - name: BATCH_DB_APIKEY
              valueFrom:
                secretKeyRef:
                  name: "{{ include "fhir.database.secretName" $ }}"
                  key: "{{ include "fhir.database.apiKeySecretKey" $ }}"
```

File: charts/ibm-fhir-server/templates/schematool.yaml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+              value: "{{ include "fhir.database.name" $ }}"
+            - name: FHIR_DB_USER
+              value: "{{ include "fhir.database.user" $ }}"
+            - name: FHIR_DB_SCHEMA
+              value: {{ .Values.db.schema }}
+            - name: FHIR_DB_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: "{{ include "fhir.database.secretName" $ }}"
+                  key: "{{ include "fhir.database.secretKey" $ }}"
```
Comment (michaelwschroeder) [inline charts/ibm-fhir-server/templates/schematool.yaml:81]:
```suggestion
                  key: "{{ include "fhir.database.passwordSecretKey" $ }}"
```
per comment in _helpers.tpl

File: charts/ibm-fhir-server/templates/schematool.yaml [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+metadata:
+  name: "{{ template "fhir.fullname" . }}-schematool"
+  labels:
+    app: {{ template "fhir.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
+    "helm.sh/hook": post-install
+    "helm.sh/hook-delete-policy": hook-succeeded
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/templates/schematool.yaml:13]: Do we need this to be a helm hook still?  Can we just run it in parallel to the deployment and remove these lines?

**Comment by lmsurpre** [comment]
@chgl I spoke with @michaelwschroeder and we definitely want these changes.
we think the benefit of having a turnkey deploy outweighs any negatives on startup/migration.
and we think it will get even better if we can get to https://github.com/IBM/FHIR/issues/2756
let us know if you need any help addressing the feedback...if needed we could do a PR to your PR :-)

**Comment by chgl** [comment]
@lmsurpre thank you! Give me a few days to rebase and address the great reviews!

File: charts/ibm-fhir-server/values.yaml (hunk 2 above)
Comment (chgl) [reply charts/ibm-fhir-server/values.yaml:62]: sounds good! This was copied over from the comments over at https://github.com/bitnami/charts/blob/master/bitnami/postgresql/values.yaml#L220

File: charts/ibm-fhir-server/templates/deployment.yaml (hunk 3 above)
Comment (chgl) [reply charts/ibm-fhir-server/templates/deployment.yaml:33]: good idea! The `pg_isready` command should be part of the db image. Should we differentiate between having `postgresql.enabled=true` and other cases? - even if `enabled=false`, once may still override the image used by the init containers.

File: charts/ibm-fhir-server/templates/_helpers.tpl [hunk 10]
```diff
@@ -35,3 +35,73 @@ Return the appropriate apiVersion for ingress.
...
+*/}}
+{{- define "fhir.database.secretKey" -}}
+{{- if (.Values.db.dbSecret) -}}
+    {{- printf "%s" .Values.db.passwordSecretKey -}}
+{{- else if .Values.postgresql.enabled }}
+    {{- printf "postgresql-password" -}}
+{{- else }}
+    {{- printf "password" -}}
+{{- end -}}
+{{- end -}}
```
[the remark this answers is not in the document]
Comment (chgl) [reply charts/ibm-fhir-server/templates/_helpers.tpl:107]: Ah, thank you! Applying it now

File: charts/ibm-fhir-server/templates/deployment.yaml (hunk 5 above)
Comment (chgl) [reply charts/ibm-fhir-server/templates/deployment.yaml:120]: The API Key is only relevant for DB2-based installations, right? - I think in that case it may be a bit more complicated, if `postgresql.enabled=true`, then `fhir.database.secretName` will resolve to `RELEASE-NAME-postgresql`, which doesn't contain an `apiKey` key. I've now went ahead and disabled setting `FHIR_DB_APIKEY` if `.Values.db.type != postgres`.

File: charts/ibm-fhir-server/Chart.yaml (hunk 4 above)
Comment (chgl) [reply charts/ibm-fhir-server/Chart.yaml:10]: easy enough, I've set:

```yaml
postgresql:
  postgresqlExtendedConf:
    maxPreparedTransactions: 24
```

which creates an `override.conf` inside the PG container:

```
max_prepared_transactions = 24
```

Can you recommend a reasonable default value?

File: charts/ibm-fhir-server/templates/schematool.yaml (hunk 9 above)
Comment (chgl) [reply charts/ibm-fhir-server/templates/schematool.yaml]: Not really, no. This is a post-install hook which is run in parallel to applying the other resources. The upside is that the `helm.sh/hook-delete-policy` automatically deletes the pod and job resource. The downsides are that this also gets rids of the logs and the `helm install` command is blocked until the job completes successfully. I've removed it for now.

**Comment by chgl** [comment]
This should address all reviews. I've added a small `helm test` which accesses the `/Patient` endpoint just to increase the confidence of it actually working. If we add https://github.com/helm/chart-testing/blob/main/doc/ct_install.md, this test can be run as part of the CI as well.

File: charts/ibm-fhir-server/values.yaml [hunk 11]
```diff
@@ -74,8 +104,13 @@ security:
   smartEnabled: false
   # -- OAuth 2.0 scopes to advertise from the server
   # @default -- openid, profile, fhirUser, launch/patient, offline_access, and a set of patient/<resource>.read scopes for a number of resource types.
-  smartScopes: [
-      "openid", "profile", "fhirUser", "launch/patient", "offline_access",
+  smartScopes:
```
Comment (chgl) [inline charts/ibm-fhir-server/values.yaml:107]: Argh, this is https://prettier.io/ auto-formatting the .yaml. ... should we add prettier to the CI to make sure the YAMLs (and Markdowns) are consistently formatted?

File: charts/ibm-fhir-server/templates/_helpers.tpl [hunk 12]
```diff
@@ -35,3 +35,88 @@ Return the appropriate apiVersion for ingress.
...
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+NOTE: we should be able to replace this approach once https://github.com/helm/helm/pull/9957 is available in Helm
+*/}}
+{{- define "fhir.postgresql.fullname" -}}
+{{- $name := default "postgresql" .Values.postgresql.nameOverride -}}
+{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Get the PostgreSQL server's hostname
```
[the remark this answers is not in the document]
Comment (chgl) [reply charts/ibm-fhir-server/templates/_helpers.tpl:50]: Good point, thanks!

File: charts/ibm-fhir-server/values.yaml (hunk 11 above)
Comment (lmsurpre) [reply charts/ibm-fhir-server/values.yaml:107]: no strong preference on this one.  can it be a pre-commit hook like `helm-docs`?

Comment (chgl) [reply charts/ibm-fhir-server/values.yaml:107]: It should work: https://prettier.io/docs/en/precommit.html#option-3-pre-commithttpsgithubcompre-commitpre-commit

**Review by michaelwschroeder — APPROVED** [review]

**Review by lmsurpre — APPROVED** [review]
I checked out the branch and ran it against a test cluster.  The postgresql statefulset came up in a couple minutes and then the fhir server pods came up.  Looks good, thanks for the great contribution @chgl !

File: charts/ibm-fhir-server/templates/tests/test-api.yaml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    heritage: {{ .Release.Service }}
+  annotations:
+    "helm.sh/hook": test
+spec:
+  restartPolicy: Never
+  containers:
+    - name: get-patient
+      image: docker.io/curlimages/curl:7.78.0
+      command: ["/bin/sh", "-exc"]
+      args: ["curl -vvv --insecure --fail -u ${FHIR_USER_NAME}:${FHIR_USER_PASSWORD} https://${FHIR_SERVER_HOST}:9443/fhir-server/api/v4/Patient"]
```
Comment (lmsurpre) [inline charts/ibm-fhir-server/templates/tests/test-api.yaml:18]: note that this will fail for auth models other than basic auth.  i recently updated the readinessProbe logic for something similar with the `$healthcheck` endpoint.  we need to work on a story for configuring OAuth / "SMART-on-FHIR" auth, so this can be addressed as part of that.
